libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,467 @@
+{
+    // calculate age of all items in the _database
+
+    if (_database.expiry_seconds == 0) {
+        // zero means never expire. This is not normal behavior but perhaps you could send a static
+        // environment once that you don't want to have to constantly update
+        return;
+
+    } else if (_database.expiry_seconds < 0) {
+        // invalid param, reset to default
+        _database.expiry_seconds.set_and_notify(AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT);
+    }
+
+    const uint32_t now_ms = AP_HAL::millis();
+    const uint32_t expiry_ms = (uint32_t)_database.expiry_seconds * 1000;
+    uint32_t index = 0;
+    while (index < _database.count) {
+        if (now_ms - _database.items[index].timestamp > expiry_ms) {
+            database_item_remove(index);
+        } else {
+            index++;
+        }
+    }
+}
+
+void AP_OADatabase::database_item_remove(const uint32_t index)
+{
+    //::printf("DB::database_item_remove()\n");
+
+    if (index >= _database.count || _database.count == 0) {
+        // index out of range
+        return;
+    }
+
+    _database.stats.dropped++;
+
+    // invalidate the location so that a UI will redraw it
+    _database.items[index] = _database.invalid_item;
+
+    database_item_send(index);
+
+    _database.count--;
+    if (_database.count == 0) {
+        return;
+    }
+
+    if (index != _database.count) {
+        // copy last object in array over expired object
+        _database.items[index] = _database.items[_database.count];
+    }
+}
+
+void AP_OADatabase::database_item_add(const OA_DbItem &item)
+{
+    if (_database.count >= _database.size) {
+        ::printf("DB::database_item_add() failed - FULL. Size: %d, filter=%.2fm\n",  _database.size,  _database.filter_m);
+        return;
+    }
+    _database.items[_database.count] = item;
+    _database.count++;
+    database_item_send(_database.count-1);
+}
+
+void AP_OADatabase::database_item_refresh(const uint32_t index, const uint32_t timestamp)
+{
+    if (index >= _database.count) {
+        // index out of range
+        return;
+    }
+
+    // update timestamp on close object so it sticks around longer
+    _database.items[index].timestamp = timestamp;
+    database_item_send(index);
+}
+
+// push a location into the database
+void AP_OADatabase::queue_push(const Location &loc, const uint32_t timestamp)
+{
+    if (!_initialised) {
+        //::printf("ni\n");
+        return;
+    }
+    //::printf("dbi\n");
+
+    const OA_DbItem item = {loc, timestamp};
+
+    WITH_SEMAPHORE(_queue.sem);
+
+    if (_queue.items->push(item) == false && (_queue.size < _queue.size_max)) {
+        // we failed to push, probably because it's full. If we're not at max size let's grow the queue size.
+        // we can't immediately deinit the queue because the background thread may be in the middle of emptying it
+        // so the defer sets a flag to let the background thread do the actual deinit
+        deinit_queue(true);
```

Comment:
If you deinit the queue from anywhere other then the avoidance thread/update enviroment then you can get a nullptr dereference within Avoidance update.

Reply:
deinit_queue is gone now so marking as resolved.